developer/src/kmc-ldml/src/util/serialize.ts

```diff
@@ -0,0 +1,273 @@
+/*
+ * Keyman is copyright (C) SIL Global. MIT License.
+ *
+ * This module contains routines for serializing from a KMXPlus file into XML.
+ */
+
+import { KMXPlus } from "@keymanapp/common-types";
+import { KeymanXMLWriter, LDMLKeyboard } from "@keymanapp/developer-utils";
+import { constants } from "@keymanapp/ldml-keyboard-constants";
+import { KeysCompiler } from "../compiler/keys.js";
+
+/**
+ * Serialize a KMXPlusFile back to XML.
+ * This is implemented for the LDML editor to be able to mutate LDML (XML) content by:
+ *  1. reading the original XML
+ *  2. compiling to KMXPlusFile
+ *  3. modifying the KMXPlusFile
+ *  4. serializing back to XML
+ *
+ * There are limitations:
+ *  - TODO-LDML-EDITOR: Transforms would not be serialized properly, due to
+ *    regex munging around markers and such.
+ *
+ *    To work around this, fields with underscores such as _from and _to are added to
+ *    the KMXPlusFile classes. These provide hints as to what the output XML should be,
+ *    and are populated by the tran compiler.
+ *
+ *  - TODO-LDML-EDITOR: Comments, whitespace, etc. are not preserved by this
+ *    approach. Updates to the XML parsing will support this, see #10622.
```

Comment:
https://github.com/keymanapp/keyman/issues/10622#issuecomment-2611405167 I think the goals I outlined here cover what we need for ldml-editor?

This does mean that we need to be able to round-trip transforms with a 'release build'

Reply:
This code does round trip, because the tran compiler sets _from and _to, etc.   The e2e test verifies this.
If the transform were mutated, _from/_to would need to be updated also.

Reply:
Yep, my preference is that we don't have two copies of the data because source of truth becomes muddy but I can appreciate that in some cases we may need to. We should clearly document which is the 'source of truth' in those circumstances, so what we are saying here is _from/_to is the source of truth and needs to be included even in 'release' builds of keyboards to .kmx -- otherwise lossless data roundtrip would not be possible.

(And thus, also, we need to clear on what is 'lossable' -- whitespace, element/attribute order, comments, etc., right?)

Reply:
For unchanged items, I wouldn't make anything lossable. If you didn't mutate that element, its lines in the file shouldn't change.

Reply:
I think we might be talking at cross purposes. For release builds of keyboards, we don't need to maintain whitespace from XML. We have different requirements for different contexts I think:

1. Release build -- eliminate all unnecessary data (comments formatting etc)
2. Debug build -- maintain a link to source xml -- at least line numbers of elements
3. Editing -- maintain a perfect view of the source xml

Is that right?

(We should maybe take this conversation out of the merged PR comment)